tempesta_fw/t/unit/bomber.c

```diff
@@ -0,0 +1,485 @@
+
+	msg.ptr = s;
+	msg.skb = NULL;
+	msg.len = strlen(s);
+	msg.flags = 0;
+
+	req = tfw_http_msg_create(&it, Conn_Clnt, msg.len);
+	tfw_http_msg_write(&it, req, &msg);
+	local_bh_disable();
+	ss_send(desc->sk, &req->msg.skb_list, false);
+	local_bh_enable();
+	tfw_http_msg_free(req);
+
+	atomic_inc(&tfw_bmb_request_nsend);
+}
+
+static int
+tfw_bmb_worker(void *data)
+{
+	int threadn = (int)(long)data;
+	uint64_t time_max;
+	int nattempt, k, i, j;
+
+	for (k = 0; k < niter; k++)
+	{
+		for (i = 0; i < nconnects; i++) {
+			tfw_bmb_connect(threadn * nconnects + i);
+		}
+
+		set_freezable();
+		time_max = (uint64_t)get_seconds() + TFW_BMB_WAIT_MAX;
+		nattempt = atomic_read(&tfw_bmb_conn_nattempt[threadn]);
+		do {
+			int nerror, ncompl;
+
+			nerror = atomic_read(&tfw_bmb_conn_nerror[threadn]);
+			ncompl = atomic_read(&tfw_bmb_conn_ncomplete[threadn]);
+			if (ncompl + nerror == nattempt) {
+				break;
+			}
+			wait_event_freezable_timeout(tfw_bmb_conn_wq,
+						     kthread_should_stop(),
+						     TFW_BMB_WAIT_INTVL);
+			if ((uint64_t)get_seconds() > time_max) {
+				TFW_ERR("%s exceeded maximum wait time of \
+					%d sec\n",
+					"worker", TFW_BMB_WAIT_MAX);
+				break;
+			}
+		} while (!kthread_should_stop());
+
+		for (i = 0; i < nconnects; i++) {
+			for (j = 0; j < nmessages; j++) {
+				if (tfw_bmb_desc[threadn][i].sk) {
+					tfw_bmb_msg_send(threadn, i);
+				}
+			}
+		}
+
+		tfw_bmb_release_sockets(threadn);
+	}
+
+	tfw_bmb_tasks[threadn] = NULL;
+	atomic_dec(&tfw_bmb_nthread);
+	wake_up(&tfw_bmb_task_wq);
+
+	do_gettimeofday(&tve);
+	return 0;
+}
+
+static void
+tfw_bmb_stop_threads(void)
+{
+	int i;
+
+	for (i = 0; i < nthreads; i++) {
+		if (tfw_bmb_tasks[i]) {
+			kthread_stop(tfw_bmb_tasks[i]);
+			tfw_bmb_tasks[i] = NULL;
+		}
+	}
+}
+
+static void
+tfw_bmb_report(void)
+{
+	int i, nattempt, ncomplete, nerror;
+
+	nattempt = ncomplete = nerror = 0;
+	for (i = 0; i < nthreads; i++)
+	{
+		nattempt += atomic_read(&tfw_bmb_conn_nattempt[i]);
+		ncomplete += atomic_read(&tfw_bmb_conn_ncomplete[i]);
+		nerror += atomic_read(&tfw_bmb_conn_nerror[i]);
+	}
+
+	printk("Initiated %d connects\n", nconnects * niter * nthreads);
+	printk("Of those %d connects initiated successfully\n",
+		nattempt);
+	printk("Of those %d connections were established successfully\n",
+		ncomplete);
+	printk("and %d connections completed with error\n",
+		nerror);
+	printk("and %d requests sent\n",
+		atomic_read(&tfw_bmb_request_nsend));
+	printk("Total time: %ld usec\n", (tve.tv_sec - tvs.tv_sec) * 1000000 +
+					 (tve.tv_usec - tvs.tv_usec));
+}
+
+static int __init
+tfw_bmb_init(void)
+{
+	int i, j, ret = 0;
+
+	fuzz_set_only_valid_gen(true);
+
+	if (tfw_addr_pton(&TFW_STR_FROM(server), &tfw_bmb_server_address)) {
+		TFW_ERR("Unable to parse server's address: %s", server);
+		return -EINVAL;
+	}
+	TFW_DBG("Started bomber module, server's address is %s\n", server);
+
+	tfw_bmb_desc = kmalloc(nthreads * sizeof(tfw_bmb_desc_t *), GFP_KERNEL);
```

Comment:
You do several allocations of ```nthreads``` of different structures. The code could be simpler if you join the structures and do only one allocation for one array of the merged structures.